File: assets/lib/python/lmntal.py
from __future__ import annotations
import random
from typing import (
    Any,
    Generator,
    List,
)


class Atom(object):
    name: str
    pass


class NormalAtom(Atom):
    arity: int
    data: Any
    args: List[Atom]
    removed: bool = False

    def __init__(self, name: str, arity: int):
        self.name = name
        self.arity = arity
        self.data = None
        self.args = [None] * arity
        self.removed = False

    def __str__(self):
        if isinstance(self.data, int) or isinstance(self.data, float):
            return f"{self.data}"
        return f"{self.name}"

    def at(self, index: int) -> Atom:
        return self.args[index]

    def remove_at(self, index: int):
        self.args[index].removed = True
        self.args[index] = None

    def is_plain(self) -> bool:
        return self.data is None

    def is_int(self) -> bool:
        return self.data is not None and isinstance(self.data, int)

    def set_int(self, data: int):
        self.data = data

    def get_int(self) -> int:
        return self.data

    def is_float(self) -> bool:
        return self.data is not None and isinstance(self.data, float)

    def is_str(self) -> bool:
        return self.data is not None and isinstance(self.data, str)


class Hyperlink(Atom):
    args: set[NormalAtom]

    def __init__(self, name: str):
        self.name = name
        self.args = set()

    def add(self, atom: NormalAtom, index: int):
        self.args.add(atom)
        atom.args[index] = self

    def remove(self, atom: NormalAtom, index: int):
        self.args.remove(atom)
        atom.args[index] = None


class AtomStore:
    store: dict[int, list[NormalAtom]]
    hyperlinks: list[Hyperlink]

    def __init__(self):
        self.store = {}
        self.hyperlinks = []

    def create(self, name: str, arity: int) -> NormalAtom:
        if arity not in self.store:
            self.store[arity] = list()
        for atom in self.store[arity]:
            # reuse removed atom
            if atom.name == name and atom.arity == arity and atom.removed:
                atom.removed = False
                return atom
        atom = NormalAtom(name, arity)
        self.store[arity].append(atom)
        return atom

    def create_hyperlink(self, name: str) -> Hyperlink:
        hl = Hyperlink(name)
        self.hyperlinks.append(hl)
        return hl

    def remove(self, atom: NormalAtom):
        atom.removed = True

    def find(self, name: int, arity: int) -> Generator[NormalAtom, None, None]:
        if arity not in self.store:
            return
        for atom in self.store[arity]:
            if atom.removed:
                continue
            if atom.name == name and atom.arity == arity:
                yield atom

    def dump(self) -> str:
        visited = set()
        res = ""
        for _, atoms in self.store.items():
            for atom in atoms:
                if not atom.is_plain or atom.removed or (atom in visited):
                    continue
                res += self.__dfs_dump(atom, visited, res)
        return res

    def __dfs_dump(self, atom: Atom, visited: set[NormalAtom], string: str) -> str:
        visited.add(atom)
        string += str(atom)
        string += "("
        count = 0
        for i in range(atom.arity):
            atom2 = atom.at(i)
            if isinstance(atom2, Hyperlink):
                string += atom2.name
                string += ","
                count += 1
            elif atom2 is not None and atom2 not in visited:
                string = self.__dfs_dump(atom2, visited, string)
                string += ","
                count += 1
        string = string[:-1]
        if count != 0:
            string += ")"
        return string


atom_list: AtomStore = AtomStore()


def create_atom(name: int, arity: int) -> NormalAtom:
    return atom_list.create(name, arity)


def create_hyperlink(name: str) -> Hyperlink:
    return atom_list.create_hyperlink(name)


def remove_atom(atom: NormalAtom):
    atom_list.remove(atom)


def find_atom(name: int, arity: int) -> Generator[NormalAtom, None, None]:
    return atom_list.find(name, arity)


def link(atom: NormalAtom, index: int, atom2: NormalAtom, index2: int):
    atom.args[index] = atom2
    atom2.args[index2] = atom


def relink(atom: NormalAtom, index: int, atom2: NormalAtom, index2: int):
    atom3 = atom.args[index]
    index3 = 0
    for i in range(atom3.arity):
        if atom3.args[i] == atom:
            index3 = i
            break
    link(atom3, index3, atom2, index2)


def get_hyperlink_at_port(atom: NormalAtom, index: int) -> Hyperlink:
    atom2 = atom.at(index)
    if isinstance(atom2, Hyperlink):
        return atom2
    return None


def equals(*atoms: Atom) -> bool:
    for i in range(len(atoms) - 1):
        if atoms[i] != atoms[i + 1]:
            return False
    return True


def dump_atoms():
    print(atom_list.dump())

File: assets/lib/java/lmntal.java
import java.util.*;
import java.util.stream.Collectors;

interface Rule {
    boolean apply();
}

enum DataType {
    PLAIN,
    INT,
    FLOAT,
    CHAR,
    STRING,
}

abstract class Atom {
    protected String name;

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public abstract int getArity();

    public void set(int index, Atom atom) {
        throw new UnsupportedOperationException();
    }

    public Atom at(int index) {
        throw new UnsupportedOperationException();
    }

    public boolean isInt() {
        return false;
    }

    public void setInt(long value) {
        throw new UnsupportedOperationException();
    }

    public long getInt() {
        throw new UnsupportedOperationException();
    }

    public void setFloat(double value) {
        throw new UnsupportedOperationException();
    }

    public void setChar(char value) {
        throw new UnsupportedOperationException();
    }
}

final class Hyperlink extends Atom {
    private final HashSet<NormalAtom> atoms;

    public Hyperlink(String name) {
        this.atoms = new HashSet<>();
        this.name = name;
    }

    public void add(NormalAtom atom, int pos) {
        atoms.add(atom);
        atom.set(pos, this);
    }

    public void remove(NormalAtom atom, int pos) {
        atoms.remove(atom);
        atom.set(pos, null);
    }

    @Override
    public int getArity() {
        return 1;
    }
}

final class NormalAtom extends Atom {
    private final Atom[] args;
    private Object data;
    private DataType type;
    private boolean isRemoved;

    public NormalAtom(String name, int arity) {
        this.name = name;
        this.type = DataType.PLAIN;
        this.args = new Atom[arity];
    }

    public static boolean equals(NormalAtom... atoms) {
        for (int i = 0; i < atoms.length - 1; i++) {
            if (atoms[i] != atoms[i + 1]) {
                return false;
            }
        }
        return true;
    }

    public boolean isRemoved() {
        return isRemoved;
    }

    public void remove() {
        isRemoved = true;
    }

    public void restore() {
        isRemoved = false;
    }

    public NormalAtom getAtomAtPort(int port, String name, int arity) {
        var atom = at(port);
        if (atom == null) {
            return null;
        }
        if (atom.getName().equals(name) && atom.getArity() == arity) {
            return (NormalAtom) atom;
        }
        return null;
    }


    public Atom at(int index) {
        return args[index];
    }

    public void set(int index, Atom atom) {
        args[index] = atom;
    }

    public void set(int index, NormalAtom atom) {
        args[index] = atom;
    }

    public long getInt() {
        return (long) data;
    }

    public double getFloat() {
        return (double) data;
    }

    public char getChar() {
        return (char) data;
    }

    public String getString() {
        return (String) data;
    }

    public boolean isPlain() {
        return type == DataType.PLAIN;
    }

    public boolean isInt() {
        return type == DataType.INT;
    }

    public void setInt(long value) {
        this.data = value;
        this.type = DataType.INT;
    }

    public boolean isFloat() {
        return type == DataType.FLOAT;
    }

    public void setFloat(double value) {
        this.data = value;
        this.type = DataType.FLOAT;
    }

    public boolean isChar() {
        return type == DataType.CHAR;
    }

    public void setChar(char value) {
        this.data = value;
        this.type = DataType.CHAR;
    }

    public boolean isString() {
        return type == DataType.STRING;
    }

    public void setString(String value) {
        this.data = value;
        this.type = DataType.STRING;
    }

    public int getArity() {
        return args.length;
    }

    public void removeAt(int index) {
        AtomStore.INSTANCE.removeAtom((NormalAtom) args[index]);
        args[index] = null;
    }

    @Override
    public String toString() {
        return switch (type) {
            case INT -> Long.toString(getInt());
            case FLOAT -> Double.toString(getFloat());
            default -> name;
        };
    }
}

class AtomStore {
    public static AtomStore INSTANCE = new AtomStore();
    private final HashMap<Integer, List<NormalAtom>> atoms;
    private final HashMap<Integer, Queue<NormalAtom>> reusableAtoms;
    private final HashSet<Hyperlink> hyperlinks;

    public AtomStore() {
        this.atoms = new HashMap<>();
        this.hyperlinks = new HashSet<>();
        this.reusableAtoms = new HashMap<>();
    }

    public NormalAtom createAtom(String name, int arity) {
        if (reusableAtoms.containsKey(arity)) {
            var queue = reusableAtoms.get(arity);
            if (!queue.isEmpty()) {
                var atom = queue.poll();
                atom.setName(name);
                atom.restore();
                return atom;
            }
        }
        var atom = new NormalAtom(name, arity);
        if (!atoms.containsKey(arity)) {
            atoms.put(arity, new ArrayList<>());
        }
        atoms.get(arity).add(atom);
        return atom;
    }

    public Hyperlink createHyperlink(String name) {
        var hl = new Hyperlink(name);
        hyperlinks.add(hl);
        return hl;
    }

    public void removeAtom(NormalAtom atom) {
        if (!reusableAtoms.containsKey(atom.getArity())) {
            reusableAtoms.put(atom.getArity(), new LinkedList<>());
        }
        atom.remove();
        var queue = reusableAtoms.get(atom.getArity());
        queue.add(atom);
    }

    public Iterable<NormalAtom> findAtom(String name, int arity) {
        if (!atoms.containsKey(arity)) {
            return new ArrayList<>();
        }
        return atoms.get(arity)
                .stream()
                .filter(atom -> atom.getName().equals(name) && !atom.isRemoved())
                .collect(Collectors.toList());
    }

    public String dumpAtoms() {
        var sb = new StringBuilder();
        var visited = new HashSet<NormalAtom>();
        for (var atoms_by_arity : atoms.values()) {
            for (var atom : atoms_by_arity) {
                if (!atom.isPlain() || atom.isRemoved() || visited.contains(atom)) {
                    continue;
                }
                dfsDump(atom, sb, visited);
            }
        }
        return sb.toString();
    }

    private void dfsDump(NormalAtom cur, StringBuilder sb, HashSet<NormalAtom> visited) {
        visited.add(cur);
        sb.append(cur.toString());
        sb.append("(");
        var count = 0;
        for (int i = 0; i < cur.getArity(); i++) {
            var atom = cur.at(i);
            if (atom instanceof Hyperlink) {
                sb.append(atom.name);
                sb.append(",");
                count++;
            } else if (atom instanceof NormalAtom) {
                if (!visited.contains(atom)) {
                    dfsDump((NormalAtom) atom, sb, visited);
                    sb.append(",");
                    count++;
                }
            }
        }
        sb.deleteCharAt(sb.length() - 1);
        if (count != 0) {
            sb.append(")");
        }
    }
}

